### PR TITLE
Make time in milliseconds 64-bits

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

Allow for longer session ticket lives.

For zd#14443

# Testing

Standard

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Review by dgarske — CHANGES_REQUESTED** [review]
Clang is unhappy:

```
Testing: ./configure --enable-all
Running: make check
make[2]: warning: -j5 forced in submake: resetting jobserver mode.
src/tls13.c:8936:16: error: implicit conversion loses integer precision: 'sword64' (aka 'long') to 'int' [-Werror,-Wshorten-64-to-32]
        return now;
        ~~~~~~ ^~~
src/tls13.c:8942:41: error: implicit conversion loses integer precision: 'long' to 'word32' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    ssl->session->ticketSeen      = now / 1000;
                                  ~ ~~~~^~~~~~
```

**Review by dgarske — CHANGES_REQUESTED** [review]
Any concern about this breaking compatibility with some embedded systems? Can we assume `sword64` is always available? Perhaps this 64-bit millisecond support would be better wrapped behind a build macro? Can you also document the 32-bit `TLS13_MAX_TIME_AGE` = 7 days limit.

**Comment by SparkiDev** [comment]
Defining WOLFSSL_32BIT_MILLI_TIME enables the old code.
We already use an sword64 with the 32-bit case but I don't want to completely break embedded customers that have implemented TimeNowInMilliseconds().

File: src/tls.c [hunk 1]
```diff
@@ -11529,10 +11534,23 @@ int TLSX_PopulateExtensions(WOLFSSL* ssl, byte isServer)
...
-                                            ssl->options.cipherSuite0,
-                                            ssl->options.cipherSuite, 1,
-                                            NULL);
+                    milli, ssl->specs.mac_algorithm, ssl->options.cipherSuite0,
+                    ssl->options.cipherSuite, 1, NULL);
+            #else
+                seen = (sword64)sess->ticketSeen * 1000 + sess->ticketSeenMilli;
+                now = TimeNowInMilliseconds();
+                if (now < seen)
+                    milli = (0xFFFFFFFFFFFFFFFFU - seen) + 1 + now;
```
Comment (lucsegers) [inline src/tls.c:11543]: Is this still correct? this was for a wrap around of 32 bit unsigned value (which was after 1.6 months).
We are now dealing with a signed value, which in practice will wrap around after 292471208,677536 years.
Theoretically this still could happen if for some reason your if TimeNowInMilliseconds is offsetted with some huge number for some reason. But then this 0xFFFFFFFFFFFFFFFFU seems wrong to me: the U is for unsigned. Not sure how to wrap around for a signed value. A sword 64 time should never overflow in my opinion...

File: src/tls13.c [hunk 2]
```diff
@@ -8933,23 +9242,34 @@ static int DoTls13NewSessionTicket(WOLFSSL* ssl, const byte* input,
         return ret;
     *inOutIdx += length;
 
+#ifdef WOLFSSL_32BIT_MILLI_TIME
     now = TimeNowInMilliseconds();
     if (now == (word32)GETTIME_ERROR)
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply src/tls13.c:9247]: A return of 0 is an indication of error for LowResTimer().
Changed TimeNowInMilli() to have 0 as error return value.

File: src/tls.c (hunk 1 above)
Comment (SparkiDev) [reply src/tls.c:11543]: Removed.

File: wolfssl/internal.h [hunk 3]
```diff
@@ -3708,14 +3718,19 @@ struct WOLFSSL_SESSION {
     word16             namedGroup;
 #endif
 #if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
-    #ifdef WOLFSSL_TLS13
+#ifdef WOLFSSL_TLS13
+#ifdef WOLFSSL_32BIT_MILLI_TIME
     word32             ticketSeen;        /* Time ticket seen (ms) */
+#else
+    word32             ticketSeen;        /* Time ticket seen (s) */
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/internal.h:3725]: A concession to platforms not supporting 64-bit platforms.
if they can't support 64 bits then they will be using WOLFSSL_32BIT_MILLI_TIME now.
Changed to be word64 and not have ticketSeenMilli.

File: wolfssl/internal.h [hunk 4]
```diff
@@ -2746,6 +2753,9 @@ typedef struct InternalTicket {
     byte            suite[SUITE_LEN];      /* cipher suite when created */
     byte            msecret[SECRET_LEN];   /* master secret */
     byte            timestamp[TIMESTAMP_LEN];          /* born on */
+#if defined(WOLFSSL_TLS13) && !defined(WOLFSSL_32BIT_MILLI_TIME)
+    byte            timestampmilli[TIMESTAMP_LEN];     /* born on milli */
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/internal.h:2757]: A concession to platforms not supporting 64-bit platforms.
if they can't support 64 bits then they will be using WOLFSSL_32BIT_MILLI_TIME now.
Changed TIMESTAMP_LEN to be 8 for reasons of alignment.

**Comment by kaleb-himes** [comment]
retest this please

**Review by dgarske — APPROVED** [review]

**Pull request merged**